File: js/Day2/dive-part-2.js
const fs = require("fs");
const instructions = fs
  .readFileSync("./js/Day2/input.txt")
  .toString()
  .split("\r\n");

const reader = (instruction) => {
  const { groups } =
    /(?<cmd>forward|down|up)\s*(?<number>\+{0,1}\-{0,1}\d+)/g.exec(instruction);
  return { cmd: groups.cmd, number: Number(groups.number) };
};

let horizontal = 0;
let vertical = 0;
let aim = 0;
for (let i = 0; i < instructions.length; i++) {
  const { cmd, number } = reader(instructions[i]);
  switch (cmd) {
    case "forward":
      horizontal += number;
      vertical += number * aim;
      break;
    case "down":
      aim += number;
      break;
    case "up":
      aim -= number;
      break;
  }
}

console.log({ horizontal, vertical, multiply: horizontal * vertical });

File: rust/Day2/dive_part_2.rs
struct Submarine;
impl Submarine {
    fn read(instruction: &str) -> Result<(&str, i64), ()> {
        if let [cmd, number] = instruction.split(" ").collect::<Vec<&str>>()[..] {
            return Ok((cmd, number.parse::<i64>().unwrap()));
        }
        Err(())
    }

    pub fn dive(instructions: Vec<&str>) -> Result<(i64, i64, i64), &str> {
        let mut horizontal = 0;
        let mut depth = 0;
        let mut aim = 0;
        for i in 0..instructions.len() {
            let (cmd, number) = Submarine::read(instructions[i]).unwrap();
            match cmd {
                "forward" => {  
                    horizontal += number;
                    depth += number * aim; 
                },
                "down" => { aim += number; },
                "up" => { aim -= number; },
                _ => { return Err("HALT: instruction unknown!"); }
            }
        }
        Ok((horizontal, depth, depth * horizontal))
    }
}

fn main() {
    let instructions: Vec<&str> = include_str!("./input.txt").lines().collect();
    
    let (horizontal, depth, multiply) = Submarine::dive(instructions).unwrap();
    println!("horizontal = {:?}, depth = {:?}, multiply = {:?}", horizontal, depth, multiply);
}